### PR TITLE
[ nG ] opset1::Reshape output shape inference refactoring

### PR DESCRIPTION
**Pull request by jane-intel**
Ticket: 32018

### CODE REVIEW

**Comment by jane-intel** [comment]
Introduced changes helped propagating semi-defined shapes through the network on the ReadNetwork stage which led to failures while validating the following operations in unpredictable places, like in Concat operation. Model example failed in pre-commit. I'll make changes and validate on the wider scope of models. Details will be listed in the ticket.

File: ngraph/core/src/op/reshape.cpp [hunk 1]
```diff
@@ -54,6 +54,122 @@ namespace
...
+                input_product *= input_pshape[i];
+            }
+        else
+            input_product = Dimension::dynamic();
+
+        if (minus_one_idx != -1) // resolving -1 masked dimension
+        {
+            if (output_product.get_min_length() == 0 && output_product.get_max_length() == 0)
+            {
+                // TODO(amprocte): Decide if this is desired behavior here. (NumPy seems
```
Comment (ilyachur) [inline ngraph/core/src/op/reshape.cpp:120]: Do we still need this comment? 😄

**Review by bsledz — APPROVED** [review]
Looks good to me

**Review by ilyachur — APPROVED** [review]

File: ngraph/core/src/op/reshape.cpp (hunk 1 above)
Comment (jane-intel) [reply ngraph/core/src/op/reshape.cpp:120]: I don't like the idea of losing this information, cause I wasn't resolving this comment in current review

File: ngraph/core/src/op/reshape.cpp [hunk 2]
```diff
@@ -54,6 +54,151 @@ namespace
...
```
Comment (jane-intel) [inline ngraph/core/src/op/reshape.cpp:120]:
```suggestion
                // TODO: Decide if this is desired behavior here. (NumPy seems
```

File: ngraph/core/src/op/reshape.cpp [hunk 3]
```diff
@@ -54,6 +54,151 @@ namespace
...
+                    // we do not include dimension to output product here and won't include in input
+                    // product later because we will divide output_product by input_product. This
+                    // dimension contributes to both products equally, but in case this dimension
+                    // is dynamic and others are not we could fully define output dimension that
+                    // is masked by -1
+                }
+            }
+            else
+            {
+                // Dimension(Interval) is a private ctor, so we use public one here
```
Comment (jane-intel) [inline ngraph/core/src/op/reshape.cpp:100]:
```suggestion
```

**Review by GlebKazantaev — APPROVED** [review]

**Pull request merged**